0.preprocess-sites/2.process-cells.py

```diff
@@ -0,0 +1,168 @@
+
+foci_dir = spot_args["output_spotdir"]
+
+sort_col = cell_args["sort_col"]
+output_basedir = cell_args["output_basedir"]
+merge_info = cell_args["merge_columns"]
+metadata_merge_left = cell_args["metadata_merge_columns"]["left"]
+metadata_merge_right = cell_args["metadata_merge_columns"]["right"]
+metadata_merge_cells_left = cell_args["metadata_merge_columns"]["cells_left"]
+
+cell_quality = CellQuality(quality_func)
+cell_category_dict = cell_quality.define_cell_quality()
+cell_category_df = pd.DataFrame(cell_category_dict, index=["Cell_Class"])
+
+# Enables feature filtering by loading the Cell Painting feature file.
+# 0.prefilter-features.py must be run first
+try:
+    all_feature_df = pd.read_csv(prefilter_file, sep="\t")
+except FileNotFoundError:
+    raise FileNotFoundError(
+        "Error",
+        f"{prefilter_file} not found.  ",
+        "Perform 0.prefilter-features.py prefilter before continuing..."
+    )
+
+sites = [x for x in os.listdir(batch_dir) if x not in ignore_files]
+
+for site in sites:
+    try:
+        print(f"Now processing {site}...")
+        output_folder = pathlib.Path(output_basedir, batch, "paint", site)
+        os.makedirs(output_folder, exist_ok=True)
+
+        meta_output_file = pathlib.Path(output_folder, f"metadata_{site}.tsv.gz")
+        count_output_file = pathlib.Path(output_folder,f"cell_counts_{site}.tsv")
+
+        compartment_dir = pathlib.Path(batch_dir, site)
+
+        #Make the compartment_csvs dictionary used to merge dfs
+        compartment_csvs = {}
+        for compartment in compartments:
+            try:
+                metadata_cols = parent_cols[compartment.lower()] + id_cols
+            except KeyError:
+                metadata_cols = id_cols
+            compartment_csvs[compartment] = load_single_cell_compartment_csv(
+                compartment_dir, compartment, metadata_cols
+            )
+
+        #Load .tsv created in 1.process-spots
+        foci_file = pathlib.Path(foci_dir, site, "full_cell_category_scores_by_guide.tsv.gz")
+        try:
+            foci_df = pd.read_csv(foci_file, sep="\t")
+        except FileNotFoundError:
+            print("""Run 1.process-spots before continuing. \n
+                  If it has been run, check that output_basedir is set correctly in
+                  the process-spots section of the config""")
+
+        #Relabel columns in foci_df to start with Metadata_Foci_
+        foci_df.columns = ["Metadata_Foci_{}".format(x) for x in foci_df.columns]
+
+    except FileNotFoundError:
+        print(f"{site} data not found")
+        continue
+
+
+    # Merge compartment csvs
+    sc_merged_df = merge_single_cell_compartments(compartment_csvs, merge_info, id_cols)
+    sc_merged_df = sc_merged_df.sort_values(by=sort_col)
+
+    # Add metadata
+    sc_merged_df = sc_merged_df.rename({"Metadata_Cells_ImageNumber": "Metadata_ImageNumber"}, axis="columns")
+    sc_merged_with_metadata_df = foci_df.merge(
```

Comment:
This merge is as written for CP074B with the merge columns extracted.
But I think something is wrong because the resulting sc_merged_with_metadata_df has a LOT of NaN.

Reply:
we expect a certain amount of NAs. The key line is [line 139](https://github.com/broadinstitute/pooled-cell-painting-profiling-recipe/pull/13/files#diff-60750dd97846eb5d02654ce784c9eb1fR139) `how="right"`.

This means that the dataframe on the right (`sc_merged_df`) (as opposed to the left, `foci_df`) is how we decide to introduce NAs. Biologically, if there are cells in `sc_merged_df` (based on the `metadta_merge_left` columns) **without foci** then all of the foci columns will have missing values (aka NAs).

[`pd.merge()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) is pretty tricky, but massively powerful.

Reply:
do we expect CP151 to have many unassigned cells?

Reply:
By unassigned cells you mean cells without foci?
I believe that is something that will be consistent between experiments because it's based on the MOI (infection rate) and that's not something that is likely to change in the immediate future.

Reply:
Note: I was misreading the df. The NaN are expected. The NaN are populating the foci metadata for cells that do not have foci in them.